vm/src/warn.rs

```diff
@@ -72,32 +72,41 @@ fn setup_context(
     let __warningregistry__ = "__warningregistry__";
     let __name__ = "__name__";
 
-    // Setup globals, filename and lineno.
-    let frame = vm.current_frame();
+    let mut f = vm.current_frame().as_deref().cloned();
 
-    //     PyThreadState *tstate = _PyThreadState_GET();
-    //     PyFrameObject *f = PyThreadState_GetFrame(tstate);
-    //     // Stack level comparisons to Python code is off by one as there is no
-    //     // warnings-related stack level to avoid.
-    //     if (stack_level <= 0 || is_internal_frame(f)) {
-    //         while (--stack_level > 0 && f != NULL) {
-    //             PyFrameObject *back = PyFrame_GetBack(f);
-    //             Py_DECREF(f);
-    //             f = back;
-    //         }
-    //     }
-    //     else {
-    //         while (--stack_level > 0 && f != NULL) {
-    //             f = next_external_frame(f);
-    //         }
-    //     }
+    // Stack level comparisons to Python code is off by one as there is no
+    // warnings-related stack level to avoid.
+    if stack_level <= 0
+        || f.clone()
+            .map(|frame| frame.is_internal_frame())
+            .unwrap_or_default()
```

Comment:
maybe redundant clone?
```suggestion
        || f.map_or(false, |frame| frame.is_internal_frame())
```